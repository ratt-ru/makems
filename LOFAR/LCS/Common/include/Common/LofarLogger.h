//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: LofarLogger.h 11806 2008-09-28 21:07:32Z loose $

#ifndef LOFAR_COMMON_LOFARLOGGER_H
#define LOFAR_COMMON_LOFARLOGGER_H

// \file
// Macro interface to the lofar logging package.

//# Include the correct set of macros, depending on the availability of the
//# log4cplus or log4cxx packages.
#if defined(HAVE_LOG4CPLUS)
# include <Common/LofarLog4Cplus.h>
#elif defined(HAVE_LOG4CXX)
# include <Common/LofarLog4Cxx.h>
#else
# include <Common/LofarLogCout.h>
#endif

#include <Common/Exceptions.h>

namespace LOFAR 
{

// \ingroup Common
// \addtogroup LofarLogger
  // @{


//# --------------------- Common definitions ---------------
//
// \name Common definitions 
// @{

// The package name of the current package. This variable should be defined in
// the file lofar_config.h which is generated by the lofarconf script. Here,
// we will make sure that LOFARLOGGER_PACKAGE is always defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

// The full package name is defined as the concatenation of the package name
// and an optional sub-package name that may be defined by the user.
#ifdef LOFARLOGGER_SUBPACKAGE
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE "." LOFARLOGGER_SUBPACKAGE
#else
# define LOFARLOGGER_FULLPACKAGE LOFARLOGGER_PACKAGE
#endif

// @}

//#------------------------ Assert and FailWhen -------------------------------
//#
//# NB: THROW is defined in package-dependant file.
//
// \name Assert and FailWhen
// \note The \c DBG version of \c ASSERT and \c FAILWHEN will only be in your
// compiled code when the (pre)compiler flag \c ENABLE_DBGASSERT is defined.
// @{

// If the condition of assert is NOT met a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERT(cond) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, "Assertion: " #cond); \
	} while(0)


// If the condition of assert is NOT met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define ASSERTSTR(cond,stream) do { \
	if (!(cond)) THROW(::LOFAR::AssertError, \
			   "Assertion: " #cond "; " << stream); \
	} while(0)

// If the condition of failwhen is met, a logrequest is sent to the logger
// <tt>\<module\>.EXCEPTION</tt> and an AssertError exception is thrown.
#define FAILWHEN(cond) do { \
	if (cond) THROW(::LOFAR::AssertError, "Failtest: " #cond); \
	} while(0)

#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

// @}
    
} // namespace LOFAR

#endif // file read before
