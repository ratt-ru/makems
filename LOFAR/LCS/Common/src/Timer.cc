//# Timer.cc: Accurate timer
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: Timer.cc 12871 2009-03-12 12:54:57Z gels $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <cassert>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iomanip>

#include <Common/PrettyUnits.h>
#include <Common/Timer.h>


using namespace std;

namespace LOFAR {


double NSTimer::CPU_speed_in_MHz = NSTimer::get_CPU_speed_in_MHz();


double NSTimer::get_CPU_speed_in_MHz()
{
    // first a few sanity checks
    assert(sizeof(int) == 4);
    assert(sizeof(long long) == 8);

#if (defined __linux__ || defined __blrts__) && \
    (defined __i386__ || defined __x86_64__ || defined __ia64__ || defined __PPC__) && \
    (defined __GNUC__ || defined __INTEL_COMPILER || defined __PATHSCALE__ || defined __xlC__)
    ifstream infile("/proc/cpuinfo");
    char     buffer[256], *colon;

    while (infile.good()) {
	infile.getline(buffer, 256);

#if defined __PPC__
	if (strcmp("cpu\t\t: 450 Blue Gene/P DD2", buffer) == 0)
	    return 850.0;

	if (strcmp("machine\t\t: Blue Gene", buffer) == 0)
	    return 700.0;

	if (strncmp("timebase", buffer, 8) == 0 && (colon = strchr(buffer, ':')) != 0)
	    return atof(colon + 2) / 1e6;
#else
	if (strncmp("cpu MHz", buffer, 7) == 0 && (colon = strchr(buffer, ':')) != 0)
	    return atof(colon + 2);
#endif
    }

    return 0.0;
#else
#warning unsupported architecture
    return 0.0;
#endif
}


double NSTimer::getElapsed() const
{
  double time = total_time / 1e6;
  if (CPU_speed_in_MHz > 0) {
    time /= CPU_speed_in_MHz;
  }
  return time;
}


ostream &NSTimer::print(ostream &str) const
{
    if (itsName.empty()) {
        str << "timer: ";
    } else {
        str << left << setw(25) << itsName << ": " << right;
    }
    if (count == 0) {
	str << "not used"; 
	if (!log_on_destruction)
	  str << endl;
    } else {
        double total = static_cast<double>(total_time);
	if (CPU_speed_in_MHz == 0) {
	    str << "avg = " << total / static_cast<double>(count);
	    str << ", total = " << total_time << " cycles";
	} else {
	    total /= 1e6 * CPU_speed_in_MHz;
	    str << "avg = " << PrettyTime(total / static_cast<double>(count))
		<< ", total = " << PrettyTime(total);
	}
	str << ", count = " << setw(9) << count;
	if (!log_on_destruction)
	  str << endl;
    }
    return str;
}


}  // end namespace LOFAR
