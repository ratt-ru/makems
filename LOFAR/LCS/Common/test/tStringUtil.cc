//#  tStringUtil.cc: test program for the string utilities class
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id: tStringUtil.cc 13875 2009-08-31 07:47:33Z diepen $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/StringUtil.h>
#include <Common/LofarLogger.h>
#include <Common/Timer.h>
#include <Common/lofar_set.h>
#include <Common/lofar_iostream.h>
#include <cmath>
#include <limits>

using namespace LOFAR;
using namespace std;

#define TOSTRING_TEST(type,min,max) \
{ \
  cout << "min(" #type ") = " << toString(min) \
       << "\t max(" #type ") = " << toString(max) << endl; \
}

#define TOSTRING_TEST2(type, val, format) \
{ \
  type x(val); \
  cout << "toString(" #type "(" #val ")," #format ") = " \
       << toString(x, format) << endl; \
}

#define	DO_COMPACT_TEST(x) \
	cout << (x) << " ==> " << compactedArrayString(x) << endl;

#define	DO_EXPAND_TEST(x) \
	cout << (x) << " ==> " << expandArrayString(x) << endl;


bool testSplit()
{
  string s(",aa,bb,,dd,");
  vector<string> vs = StringUtil::split(s,',');
  cout << "Splitting string \"" << s << "\" using \',\' as seperator ..." 
       << endl;
  for (string::size_type i = 0; i < vs.size(); i++) {
    cout << "vs[" << i << "] = \"" << vs[i] << "\"" << endl;
  }
  cout << endl;
  return true;
}


bool testCompare(bool nocase)
{
  typedef set<string, StringUtil::Compare> set_type;
  set_type 
    s(nocase ? StringUtil::Compare::NOCASE : StringUtil::Compare::NORMAL);
  s.insert("Alfons");
  s.insert("aap");
  s.insert("Bertus");
  s.insert("bakker");
  s.insert("Chris");
  s.insert("chocolade");
  if (nocase) cout << "=== Strings sorted NOCASE ===" << endl;
  else cout << "=== Strings sorted NORMAL ===" << endl;
  for (set_type::const_iterator it = s.begin(); it != s.end(); ++it) {
    cout << *it << endl;
  }
  cout << endl;
  return true;
}


bool testTrim()
{
  string r1(" 	 	 a String with leading and trailing whitespace	 ");
  string r2("a String without leading and trailing whitespace");
  string r3("1) String with(out) itemnr and arrayindex[123]");

  rtrim(r1);
  ltrim(r1);
  rtrim(r2);
  ltrim(r2);
  rtrim(r3, "0123456789[] ");
  ltrim(r3, "0123456789() ");

  cout << "Trimmed C++ strings:" << endl;
  cout << ">" << r1 << "<" << endl;
  cout << ">" << r2 << "<" << endl;
  cout << ">" << r3 << "<" << endl;

  char c1[] = { " 	 	 a String with leading and trailing whitespace	 " };
  char c2[] = { "a String without leading and trailing whitespace" };
  char c3[] = { "1) String with(out) itemnr and arrayindex[123]" };
  char* p1 = c1;
  char* p2 = c2;
  char* p3 = c3;
  int32 len1, len2, len3;

  len1 = rtrim(p1 = ltrim(p1));
  len2 = rtrim(p2 = ltrim(p2));
  p3 = ltrim(p3, "0123456789() ");
  len3 = rtrim(p3 = ltrim(p3, "0123456789() ") , 0, "0123456789[] ");

  cout << "Trimmed C strings:" << endl;
  cout << ">" << p1 << "< , len=" << len1 << endl;
  cout << ">" << p2 << "< , len=" << len2 << endl;
  cout << ">" << p3 << "< , len=" << len3 << endl;

  return true;
}


bool testSkipws()
{
  {
    string str;
    uint st = lskipws(str, 0, str.size());
    uint end = rskipws(str, st, str.size());
    ASSERT (st==0 && end==0);
  }
  {
    string str(" \t \t\t  ");
    uint st = lskipws(str, 0, str.size());
    uint end = rskipws(str, st, str.size());
    ASSERT (st==7 && end==7);
  }
  {
    string str(" \t \t\ta  ");
    uint st = lskipws(str, 0, str.size());
    uint end = rskipws(str, st, str.size());
    ASSERT (st==5 && end==6);
    ASSERT (lskipws(str, st, end) == st);
    ASSERT (lskipws(str, st, str.size()) == st);
    ASSERT (lskipws(str, end, end) == end);
    ASSERT (lskipws(str, end, str.size()) == str.size());
  }
  return true;
}


bool testCase()
{
  cout << "Lowercase and uppercase a string:" << endl;
  string s("The zip code of Hondelage in Germany is 38108");
  cout << "orginal: " << s << endl;
  cout << "lowered: " << toLower(s) << endl;
  cout << "uppered: " << toUpper(s) << endl;
  return true;
}


bool testToString()
{
  //## Conversion of fundamental arithmetic types to string ##//

  cout << "\n*** Conversion of fundamental arithmetic types to string ***\n";
  try {
    TOSTRING_TEST(bool,false,true);
    TOSTRING_TEST(char,0,127);
    TOSTRING_TEST(signed char,-128,127);
    TOSTRING_TEST(unsigned char,0,255);
    TOSTRING_TEST(short,-32768,32767);
    TOSTRING_TEST(unsigned short,0,65535);
    TOSTRING_TEST(int,-32768,32767);
    TOSTRING_TEST(unsigned int,0,65536);
    TOSTRING_TEST(long,~0x7FFFFFFFL,0x7FFFFFFFL);
    TOSTRING_TEST(unsigned long,0,0xFFFFFFFFUL);
#ifdef HAVE_LONG_LONG
    TOSTRING_TEST(long long,~0x7FFFFFFFFFFFFFFFLL,0x7FFFFFFFFFFFFFFFLL);
    TOSTRING_TEST(unsigned long long,0,0xFFFFFFFFFFFFFFFFULL);
#endif
    TOSTRING_TEST(float,numeric_limits<float>::min(),numeric_limits<float>::max());
    TOSTRING_TEST(double,numeric_limits<double>::min(), numeric_limits<double>::max());
#ifdef HAVE_LONG_DOUBLE
    TOSTRING_TEST(long double,numeric_limits<long double>::min(), numeric_limits<long double>::max());
#endif
    cout << endl;
    TOSTRING_TEST2(int, 42, "%06i");
    TOSTRING_TEST2(float, M_E, "%e");
    TOSTRING_TEST2(float, M_PI, "%8.4f");
    TOSTRING_TEST2(double, M_E, "%+08.12g");
    TOSTRING_TEST2(double, M_PI*1e12, "%+08.12g");
  } catch (Exception& e) {
    LOG_ERROR_STR(e);
    return false;
  }
  return true;
}

bool testStrTo()
{
  ASSERT (strToLong("-123") == -123);
  ASSERT (strToLong("  -123  ") == -123);
  ASSERT (strToLong("  -0123  ") == -123);
  ASSERT (strToLong(" 0x123 ") == 0x123);
  ASSERT (strToInt("-123") == -123);
  ASSERT (strToInt32("-123") == -123);
  ASSERT (strToInt16(" 124") == 124);
  ASSERT (strToUlong("1234") == 1234);
  ASSERT (strToUint("1235") == 1235);
  ASSERT (strToUint32("1236") == 1236);
  ASSERT (strToUint16("1237") == 1237);
  bool fail=false;
  try {
    strToInt32("  1234567890000  ");
  } catch (...) {
    fail = true;
  }
  ASSERT (fail);
  fail=false;
  try {
    strToUint16("  65536  ");
  } catch (...) {
    fail = true;
  }
  ASSERT (fail);
  fail=false;
  try {
    strToInt16("  6.5  ");
  } catch (...) {
    fail = true;
  }
  ASSERT (fail);
  ASSERT (strToFloat("6") == 6);
  ASSERT (strToFloat("  6.5") == 6.5);
  ASSERT (strToFloat("  65e-1  ") == 6.5);
  ASSERT (strToDouble("  6.5  ") == 6.5);
  fail=false;
  try {
    strToDouble("  6.5a  ");
  } catch (...) {
    fail = true;
  }
  ASSERT (fail);
#ifdef HAVE_LONG_LONG
  ASSERT (strToInt64 ("  -12345678901234  ") == int64(-12345678901234LL));
  ASSERT (strToUint64("   12345678901234  ") == uint64(12345678901234LL));
#endif
  return true;
}

bool testCompactArray()
{
  cout << "\n*** Testing compacting of Array-strings ***\n";
  try {
	DO_COMPACT_TEST("[ a,b,c ]");
	DO_COMPACT_TEST("[ aa01,aa06 ]");
	DO_COMPACT_TEST("[ aa01,aa02 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03,aa04 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03,aa05 ]");
	DO_COMPACT_TEST("[ aa01, aa02 ,aa03,aa05,aa06 ]");
	DO_COMPACT_TEST("[ aa01, bb02 ,aa03,aa05,aa06 ]");
	DO_COMPACT_TEST("[ aa01, aa02 ,aa0003,aa05,aa06 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03,aa05,aa06,aa7 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03,aa05,a06,aa7 ]");
	DO_COMPACT_TEST("[ aa01,aa02,aa03,aa05,aaa06,aa7 ]");

	DO_COMPACT_TEST("[22,23,24,25,30,31,33,35]");
	DO_COMPACT_TEST("[22,23,24,25,30,31,33,35,36]");
	DO_COMPACT_TEST("[22,23,24,25,30,31,33,35,36,37]");
	DO_COMPACT_TEST("[22,23,24,25,30,31,33,35,36,37,37]");
	DO_COMPACT_TEST("[22,23,24,25,30,31,33,35,36,37,40]");
	DO_COMPACT_TEST("[22,23,23,25,30,31,30]");
	DO_COMPACT_TEST("[22,23,23,25,30,31,30,31]");
	DO_COMPACT_TEST("[22,23,23,25,30,31,30,31,31]");
	DO_COMPACT_TEST("[22,23,23,23,23,23,25,30,31,30,31,31]");

	DO_COMPACT_TEST("[23..28]");
	DO_COMPACT_TEST("[22,23..28]");
	DO_COMPACT_TEST("[22,23..28,29]");
	DO_COMPACT_TEST("[21,23..28,29]");
	DO_COMPACT_TEST("[21,23..28,30]");
	DO_COMPACT_TEST("[23..28,30..32]");
	DO_COMPACT_TEST("[23..28,29..32]");
	DO_COMPACT_TEST("[23..28,29,30..32]");
	DO_COMPACT_TEST("[23..26,29,30..32]");
	DO_COMPACT_TEST("[23..26,27,30..32]");
  } catch(Exception& e) {
	LOG_ERROR_STR(e);
	return (false);
  }
  return (true);
}

bool testExpandArray()
{	
  cout << "\n*** Testing expanding of Array-strings ***\n";
  try {
        // Test expansions.
        DO_EXPAND_TEST("[3*ab]");
        DO_EXPAND_TEST("[3*(ab)]");
        DO_EXPAND_TEST("[3*'ab']");
        DO_EXPAND_TEST("[3*('ab')]");
        DO_EXPAND_TEST("[3*10.5]");
        DO_EXPAND_TEST("[3*(10.5)]");
        DO_EXPAND_TEST("[lifs0..lifs10]");
        DO_EXPAND_TEST("[lifs00..lifs010]");
        DO_EXPAND_TEST("[lifs000..lifs010]");
        DO_EXPAND_TEST("[2*lifs000..lifs010]");
        DO_EXPAND_TEST("[2*(lifs000..lifs010)]");
        DO_EXPAND_TEST("[2*([1,2])]");

        cout << "  Some more involved expansions" << endl;
        DO_EXPAND_TEST("[abc]");
        DO_EXPAND_TEST("[abc0..abc9]");
        DO_EXPAND_TEST("['abc0..abc9']");
        DO_EXPAND_TEST("[abc000..abc9]");
        DO_EXPAND_TEST("[abc000..abc9def]");
        DO_EXPAND_TEST("[abc{000..9}def]");
        DO_EXPAND_TEST("[abc${NAME}def]");
        DO_EXPAND_TEST("[(abc0009  ..  abc4, 010..8)]");
        DO_EXPAND_TEST("[/abc/../cd0..3]");
        DO_EXPAND_TEST("[3*abc]");
        DO_EXPAND_TEST("[3 * [[1,2,3],[4,5,6]], 02*'xx']");
        DO_EXPAND_TEST("[3 * [[5*1],[5*2]], 02*'xx']");
        DO_EXPAND_TEST("[2*(0,1,2,3), 2*(5;6;7), 2*10..12]");
        DO_EXPAND_TEST(" [ 3 * 2 * 10 .. 12 ] ");
        DO_EXPAND_TEST(" [3*(2*3*6;2*(12,13))]");

        cout << "  Miscellaneous tests with expandArrayString" << endl;
        DO_EXPAND_TEST(" [3*'10.5*ab']");
        DO_EXPAND_TEST(" [10.5*'ab']");
        DO_EXPAND_TEST(" [3*10.5*'ab']");
        DO_EXPAND_TEST(" [3*'ab'*2]");
        DO_EXPAND_TEST(" [3*ab*2]");
        DO_EXPAND_TEST(" [3*3*2]");
        DO_EXPAND_TEST(" [3*(3*2)]");
        DO_EXPAND_TEST(" [3*10.5*2]");
        DO_EXPAND_TEST(" [3*(10.5*2)]");
        DO_EXPAND_TEST(" [(1,2,3)]");

        cout << "  Performance test of expandArrayString" << endl;
        cout << ">>>" << endl;
        NSTimer tim2;
        tim2.start();
        string s2 = expandArrayString("[abc001..abc10,abc45..abc50]");
        tim2.stop();
        tim2.print(cout);
        cout << "<<<" << endl;
        cout << s2 << endl;

  } catch(Exception& e) {
	LOG_ERROR_STR(e);
	return (false);
  }
  return (true);
}

int main()
{
  INIT_LOGGER("tStringUtil");

  bool result = 
    testSplit()        &&
    testCompare(false) &&
    testCompare(true)  &&
    testTrim()         &&
    testSkipws()       &&
    testCase()         &&
    testToString()     &&
    testStrTo()        &&
    testCompactArray() &&
    testExpandArray();

  return (result ? 0 : 1);
}
