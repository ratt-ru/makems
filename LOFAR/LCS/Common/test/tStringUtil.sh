#!/bin/sh
./runctest.sh tStringUtil 2>&1 > tStringUtil.log
