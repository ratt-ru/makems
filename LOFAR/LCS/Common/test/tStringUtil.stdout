Splitting string ",aa,bb,,dd," using ',' as seperator ...
vs[0] = ""
vs[1] = "aa"
vs[2] = "bb"
vs[3] = ""
vs[4] = "dd"
vs[5] = ""

=== Strings sorted NORMAL ===
Alfons
Bertus
Chris
aap
bakker
chocolade

=== Strings sorted NOCASE ===
aap
Alfons
bakker
Bertus
chocolade
Chris

Trimmed C++ strings:
>a String with leading and trailing whitespace<
>a String without leading and trailing whitespace<
>String with(out) itemnr and arrayindex<
Trimmed C strings:
>a String with leading and trailing whitespace< , len=45
>a String without leading and trailing whitespace< , len=48
>String with(out) itemnr and arrayindex< , len=38
Lowercase and uppercase a string:
orginal: The zip code of Hondelage in Germany is 38108
lowered: the zip code of hondelage in germany is 38108
uppered: THE ZIP CODE OF HONDELAGE IN GERMANY IS 38108

*** Conversion of fundamental arithmetic types to string ***
min(bool) = false	 max(bool) = true
min(char) = 0	 max(char) = 127
min(signed char) = -128	 max(signed char) = 127
min(unsigned char) = 0	 max(unsigned char) = 255
min(short) = -32768	 max(short) = 32767
min(unsigned short) = 0	 max(unsigned short) = 65535
min(int) = -32768	 max(int) = 32767
min(unsigned int) = 0	 max(unsigned int) = 65536
min(long) = -2147483648	 max(long) = 2147483647
min(unsigned long) = 0	 max(unsigned long) = 4294967295
min(long long) = -9223372036854775808	 max(long long) = 9223372036854775807
min(unsigned long long) = 0	 max(unsigned long long) = 18446744073709551615
min(float) = 1.17549e-38	 max(float) = 3.40282e+38
min(double) = 2.22507e-308	 max(double) = 1.79769e+308

toString(int(42),"%06i") = 000042
toString(float(M_E),"%e") = 2.718282e+00
toString(float(M_PI),"%8.4f") =   3.1416
toString(double(M_E),"%+08.12g") = +2.71828182846
toString(double(M_PI*1e12),"%+08.12g") = +3.14159265359e+12

*** Testing compacting of Array-strings ***
[ a,b,c ] ==> [ a,b,c ]
[ aa01,aa06 ] ==> [aa01,aa06]
[ aa01,aa02 ] ==> [aa01,aa02]
[ aa01,aa02,aa03 ] ==> [aa01..aa03]
[ aa01,aa02,aa03,aa04 ] ==> [aa01..aa04]
[ aa01,aa02,aa03,aa05 ] ==> [aa01..aa03,aa05]
[ aa01, aa02 ,aa03,aa05,aa06 ] ==> [aa01..aa03,aa05,aa06]
[ aa01, bb02 ,aa03,aa05,aa06 ] ==> [ aa01, bb02 ,aa03,aa05,aa06 ]
[ aa01, aa02 ,aa0003,aa05,aa06 ] ==> [aa01..aa03,aa05,aa06]
[ aa01,aa02,aa03,aa05,aa06,aa7 ] ==> [aa01..aa03,aa05..aa07]
[ aa01,aa02,aa03,aa05,a06,aa7 ] ==> [ aa01,aa02,aa03,aa05,a06,aa7 ]
[ aa01,aa02,aa03,aa05,aaa06,aa7 ] ==> [ aa01,aa02,aa03,aa05,aaa06,aa7 ]
[22,23,24,25,30,31,33,35] ==> [22..25,30,31,33,35]
[22,23,24,25,30,31,33,35,36] ==> [22..25,30,31,33,35,36]
[22,23,24,25,30,31,33,35,36,37] ==> [22..25,30,31,33,35..37]
[22,23,24,25,30,31,33,35,36,37,37] ==> [22..25,30,31,33,35..37,37]
[22,23,24,25,30,31,33,35,36,37,40] ==> [22..25,30,31,33,35..37,40]
[22,23,23,25,30,31,30] ==> [22,23,23,25,30,31,30]
[22,23,23,25,30,31,30,31] ==> [22,23,23,25,30,31,30,31]
[22,23,23,25,30,31,30,31,31] ==> [22,23,23,25,30,31,30,31,31]
[22,23,23,23,23,23,25,30,31,30,31,31] ==> [22,23,23,23,23,23,25,30,31,30,31,31]
[23..28] ==> [23..28]
[22,23..28] ==> [22..28]
[22,23..28,29] ==> [22..29]
[21,23..28,29] ==> [21,23..29]
[21,23..28,30] ==> [21,23..28,30]
[23..28,30..32] ==> [23..28,30..32]
[23..28,29..32] ==> [23..32]
[23..28,29,30..32] ==> [23..32]
[23..26,29,30..32] ==> [23..26,29..32]
[23..26,27,30..32] ==> [23..27,30..32]

*** Testing expanding of Array-strings ***
[3*ab] ==> [ab,ab,ab]
[3*(ab)] ==> [ab,ab,ab]
[3*'ab'] ==> ['ab','ab','ab']
[3*('ab')] ==> ['ab','ab','ab']
[3*10.5] ==> [10.5,10.5,10.5]
[3*(10.5)] ==> [10.5,10.5,10.5]
[lifs0..lifs10] ==> [lifs0,lifs1,lifs2,lifs3,lifs4,lifs5,lifs6,lifs7,lifs8,lifs9,lifs10]
[lifs00..lifs010] ==> [lifs00,lifs01,lifs02,lifs03,lifs04,lifs05,lifs06,lifs07,lifs08,lifs09,lifs10]
[lifs000..lifs010] ==> [lifs000,lifs001,lifs002,lifs003,lifs004,lifs005,lifs006,lifs007,lifs008,lifs009,lifs010]
[2*lifs000..lifs010] ==> [lifs000,lifs001,lifs002,lifs003,lifs004,lifs005,lifs006,lifs007,lifs008,lifs009,lifs010,lifs000,lifs001,lifs002,lifs003,lifs004,lifs005,lifs006,lifs007,lifs008,lifs009,lifs010]
[2*(lifs000..lifs010)] ==> [lifs000,lifs001,lifs002,lifs003,lifs004,lifs005,lifs006,lifs007,lifs008,lifs009,lifs010,lifs000,lifs001,lifs002,lifs003,lifs004,lifs005,lifs006,lifs007,lifs008,lifs009,lifs010]
[2*([1,2])] ==> [[1,2],[1,2]]
  Some more involved expansions
[abc] ==> [abc]
[abc0..abc9] ==> [abc0,abc1,abc2,abc3,abc4,abc5,abc6,abc7,abc8,abc9]
['abc0..abc9'] ==> ['abc0..abc9']
[abc000..abc9] ==> [abc000,abc001,abc002,abc003,abc004,abc005,abc006,abc007,abc008,abc009]
[abc000..abc9def] ==> [abc000def,abc001def,abc002def,abc003def,abc004def,abc005def,abc006def,abc007def,abc008def,abc009def]
[abc{000..9}def] ==> [abc000def,abc001def,abc002def,abc003def,abc004def,abc005def,abc006def,abc007def,abc008def,abc009def]
[abc${NAME}def] ==> [abc${NAME}def]
[(abc0009  ..  abc4, 010..8)] ==> [(abc0009,abc0008,abc0007,abc0006,abc0005,abc0004, 010,009,008)]
[/abc/../cd0..3] ==> [/abc/../cd0,/abc/../cd1,/abc/../cd2,/abc/../cd3]
[3*abc] ==> [abc,abc,abc]
[3 * [[1,2,3],[4,5,6]], 02*'xx'] ==> [[[1,2,3],[4,5,6]],[[1,2,3],[4,5,6]],[[1,2,3],[4,5,6]], 'xx','xx']
[3 * [[5*1],[5*2]], 02*'xx'] ==> [[[1,1,1,1,1],[2,2,2,2,2]],[[1,1,1,1,1],[2,2,2,2,2]],[[1,1,1,1,1],[2,2,2,2,2]], 'xx','xx']
[2*(0,1,2,3), 2*(5;6;7), 2*10..12] ==> [0,1,2,3,0,1,2,3, 5,6,7,5,6,7, 10,11,12,10,11,12]
 [ 3 * 2 * 10 .. 12 ]  ==>  [ 10,11,12 ,10,11,12 ,10,11,12 ,10,11,12 ,10,11,12 ,10,11,12 ] 
 [3*(2*3*6;2*(12,13))] ==>  [6,6,6,6,6,6,12,13,12,13,6,6,6,6,6,6,12,13,12,13,6,6,6,6,6,6,12,13,12,13]
  Miscellaneous tests with expandArrayString
 [3*'10.5*ab'] ==>  ['10.5*ab','10.5*ab','10.5*ab']
 [10.5*'ab'] ==>  [10.5*'ab']
 [3*10.5*'ab'] ==>  [10.5*'ab',10.5*'ab',10.5*'ab']
 [3*'ab'*2] ==>  ['ab'*2,'ab'*2,'ab'*2]
 [3*ab*2] ==>  [ab*2,ab*2,ab*2]
 [3*3*2] ==>  [2,2,2,2,2,2,2,2,2]
 [3*(3*2)] ==>  [2,2,2,2,2,2,2,2,2]
 [3*10.5*2] ==>  [10.5*2,10.5*2,10.5*2]
 [3*(10.5*2)] ==>  [10.5*2,10.5*2,10.5*2]
 [(1,2,3)] ==>  [(1,2,3)]
  Performance test of expandArrayString
>>>
timer: avg = 46163, total = 46163 cycles, count =         1
<<<
[abc001,abc002,abc003,abc004,abc005,abc006,abc007,abc008,abc009,abc010,abc45,abc46,abc47,abc48,abc49,abc50]
